### PR TITLE
Bookkeeping for reachability of nodes; reporting them to Lokid

### PR DESCRIPTION
**Pull request by msgmaxim**
- Keep track of all nodes that failed the reachability test.
- Periodically test two nodes: one randomly selected and one that is known to have failed previously (this is done to give failing nodes a better chance to be recommissioned as soon as possible once they are back online).
- When a node is not reachable for longer than 120mins (to match uptime proofs) it will be reported to Lokid via a new endpoint. The same endpoint is used to report previously offending nodes as healthy.
- To be on the safe side and keep things simple, there is currently some redundancy in how many times a node's status is reported to Lokid.

### CODE REVIEW

File: httpserver/service_node.cpp [hunk 1]
```diff
@@ -286,6 +301,8 @@ void ServiceNode::bootstrap_data() {
                 if (res.error_code == SNodeError::NO_ERROR) {
                     try {
                         const block_update_t bu = parse_swarm_update(res.body);
+                        // TODO: this should be disabled in the "testnet" mode
+                        // (or changed to point to testnet seeds)
```
[the remark this answers is not in the document]
Comment (msgmaxim) [reply httpserver/service_node.cpp]: Yeah, I know 👍

File: common/include/loki_common.h [hunk 2]
```diff
@@ -9,20 +9,24 @@
...
 
     // our 32 byte pub keys should always be 52 bytes long in base32z
     static constexpr size_t BASE_LEN = 52;
 
   private:
     uint16_t port_;
-    std::string sn_address_; // Snode address
-    std::string pub_key_;
+    std::string sn_address_; // Snode address (pubkey plus .snode)
+    std::string pub_key_;    // base32z
```
[the remark this answers is not in the document]
Comment (msgmaxim) [reply common/include/loki_common.h]: Sure, renamed

File: httpserver/service_node.cpp [hunk 3]
```diff
@@ -656,11 +673,12 @@ void ServiceNode::swarm_timer_tick() {
     fields["height"] = true;
     fields["block_hash"] = true;
     fields["hardfork"] = true;
+    fields["funded"] = true;
 
     params["fields"] = fields;
 
     /// TODO: include decommissioned
```
[the remark this answers is not in the document]
Comment (msgmaxim) [reply httpserver/service_node.cpp]: Yeah, removed

File: httpserver/service_node.cpp [hunk 4]
```diff
@@ -697,23 +715,59 @@ void ServiceNode::ping_peers_tick() {
...
+    // TODO: there is an edge case where SS reported some offending
+    // nodes, but then restarted, so SS won't give priority to those
+    // nodes. SS will still test them eventually (through random selection) and
+    // update Lokid, but this scenario could be made more robust.
+    const auto offline_node = reach_records_.next_to_test();
+
+    if (offline_node) {
+        const boost::optional<sn_record_t> sn =
+            swarm_->get_node_by_pk(*offline_node);
+        LOKI_LOG(debug, "No nodes offline nodes to ping test yet");
```
[the remark this answers is not in the document]
Comment (msgmaxim) [reply httpserver/service_node.cpp]: Updated

File: httpserver/service_node.cpp [hunk 5]
```diff
@@ -697,23 +715,59 @@ void ServiceNode::ping_peers_tick() {
...
+            // delete its entry from test records as irrelevant
+            reach_records_.expire(*offline_node);
+        }
+    }
 }
 
 void ServiceNode::test_reachability(const sn_record_t& sn) {
 
-    LOKI_LOG(info, "TODO: testing node {}", sn);
+    LOKI_LOG(debug, "testing node for reachability {}", sn);
```
[the remark this answers is not in the document]
Comment (msgmaxim) [reply httpserver/service_node.cpp]: Updated

File: httpserver/service_node.cpp [hunk 6]
```diff
@@ -933,6 +987,88 @@ void ServiceNode::send_blockchain_test_req(const sn_record_t& testee,
...
+                }
+            } catch (...) {
+                LOKI_LOG(error,
+                         "Could not report node status: bad json in response");
+            }
+        } else {
+            LOKI_LOG(error, "Could not report node status");
+        }
+
+        if (success) {
```
[the remark this answers is not in the document]
Comment (msgmaxim) [reply httpserver/service_node.cpp]: I like the suggestion to invert the above check, updated accordingly (although it leads to a double negative...). I still want to keep the `success` variable to make things explicit (and to protect against us adding another failure branch and forgetting to return early).

File: httpserver/swarm.cpp [hunk 7]
```diff
@@ -150,13 +150,14 @@ static all_swarms_t apply_ips(const all_swarms_t& swarms_to_keep,
 
 void Swarm::apply_swarm_changes(const all_swarms_t& new_swarms) {
 
-    all_cur_swarms_ = apply_ips(new_swarms, all_cur_swarms_);
+    all_valid_swarms_ = apply_ips(new_swarms, all_valid_swarms_);
 }
 
 void Swarm::update_state(const all_swarms_t& swarms,
+                         const std::vector<sn_record_t>& decommissioned,
```
Comment (sachaaaaa) [inline httpserver/swarm.cpp]: Not sure if I am only looking at a partial diff, but I can't see where this new `decommissioned` variable is being used in `update_state`?

Comment (msgmaxim) [reply httpserver/swarm.cpp]: That's a bug, good catch! Should be adding them to `all_funded_nodes_`, otherwise they wouldn't be tested and won't have a chance to come back online! Fixed now.

File: httpserver/reachability_testing.cpp [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+bool reachability_records_t::expire(const sn_pub_key_t& sn) {
+
+    if (offline_nodes_.erase(sn)) {
+        LOKI_LOG(warn, "    - removed entry for {}", sn);
+    }
+}
+
+void reachability_records_t::set_reported(const sn_pub_key_t& sn) {
+
+    auto it = offline_nodes_.find(sn);
```
[the remark this answers is not in the document]
Comment (msgmaxim) [reply httpserver/reachability_testing.cpp]: Updated

File: httpserver/reachability_testing.cpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+            std::chrono::duration_cast<std::chrono::seconds>(elapsed).count();
+        LOKI_LOG(info, "    - first time failed {} seconds ago", elapsed_sec);
+
+        /// TODO: Might still want to report as unreachable since this status
+        /// gets reset to `true` on Lokid restart
+        if (elapsed > UNREACH_GRACE_PERIOD && !it->second.reported) {
+            LOKI_LOG(warn, "    - will REPORT this node to Lokid!");
+            return true;
+        } else {
+            if (it->second.reported) {
```
[the remark this answers is not in the document]
Comment (msgmaxim) [reply httpserver/reachability_testing.cpp]: Sure, and I changed the order to simplify further.

File: httpserver/reachability_testing.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+} // namespace detail
+
+/// How long to wait until reporting unreachable nodes to Lokid
+constexpr std::chrono::minutes UNREACH_GRACE_PERIOD = 120min;
+
+bool reachability_records_t::record_unreachable(const sn_pub_key_t& sn) {
+
+    auto it = offline_nodes_.find(sn);
```
[the remark this answers is not in the document]
Comment (msgmaxim) [reply httpserver/reachability_testing.cpp]: Yes, you are right. Updated.

File: httpserver/reachability_testing.h [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace loki {
+
+namespace detail {
+
+/// TODO: make this class "private"?
+class reach_record_t {
+
+    // The time the node failed for the first time
+    // (and hasn't come back online)
```
[the remark this answers is not in the document]
Comment (msgmaxim) [reply httpserver/reachability_testing.h]: Yeah, updated.
